pkg/operator/controller/dns_status.go

```diff
@@ -52,34 +68,96 @@ func computeDNSStatusConditions(dns *operatorv1.DNS, clusterIP string, haveDNSDa
+	conditions = append(conditions, computeDNSAvailableCondition(oldAvailableCondition, clusterIP, haveDNSDaemonset, dnsDaemonset))
+	conditions = append(conditions, computeDNSUpgradeableCondition(oldUpgradeableCondition, dns))
 
-	return conditions
+	return conditions, retryErr
 }
 
+var (
+	// Degraded indicates the DNS controller is degraded
+	DNSControllerDegradedConditionType = "Degraded"
+
+	// DNSNoService indicates that no IP address is assigned to the DNS service
+	DNSNoService = "NoService"
+
+	// DNSNoDNSDaemonSet indicates that the DNS daemon set doesn't exist
+	DNSNoDNSDaemonSet = "NoDaemonSet"
+
+	// DNSNoDNSPodsDesired indicates that no DNS pods are desired; this could mean
+	// all nodes are tainted or unschedulable
+	DNSNoDNSPodsDesired = "NoDNSPodsDesired"
+
+	// DNSNoDNSPodsAvailable indicates that no DNS pods are available
+	DNSNoDNSPodsAvailable = "NoDNSPodsAvailable"
+
+	// DNSInvalidDNSMaxUnavailable indicates that the DNS daemonset has an invalid
+	// MaxUnavailable value configured
+	DNSInvalidDNSMaxUnavailable = "InvalidDNSMaxUnavailable"
+
+	// DNSMaxUnavailableDNSPodsExceeded indicates that the number of unavailable DNS
+	// pods is greater than the configured MaxUnavailable
+	DNSMaxUnavailableDNSPodsExceeded = "MaxUnavailableDNSPodsExceeded"
+)
+
 // computeDNSDegradedCondition computes the dns Degraded status condition
 // based on the status of clusterIP and the DNS daemonset.  The node-resolver
 // daemonset is not a part of the calculation of degraded condition.
-func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clusterIP string, haveDNSDaemonset bool, dnsDaemonset *appsv1.DaemonSet) operatorv1.OperatorCondition {
-	degradedCondition := &operatorv1.OperatorCondition{
+func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clusterIP string, haveDNSDaemonset bool, dnsDaemonset *appsv1.DaemonSet) (operatorv1.OperatorCondition, error) {
+	finalDegradedCondition := operatorv1.OperatorCondition{
 		Type: operatorv1.OperatorStatusTypeDegraded,
 	}
-	status := operatorv1.ConditionUnknown
-	degradedReasons := []string{}
+	degradedConditions := []operatorv1.OperatorCondition{}
 	messages := []string{}
+	var keepConditionUnknown bool
+	now := metav1.NewTime(clock.Now())
+
+	// These conditions are expected to be false
+	expectedConditions := []dnsCondition{
+		{
+			condition: DNSNoService,
+			status:    operatorv1.ConditionFalse,
+		},
+		{
+			condition: DNSControllerDegradedConditionType,
+			status:    operatorv1.ConditionFalse,
+		},
+		{
+			condition: DNSNoDNSDaemonSet,
+			status:    operatorv1.ConditionFalse,
+		},
+		{
+			condition: DNSNoDNSPodsDesired,
+			status:    operatorv1.ConditionFalse,
+		},
+		{
+			condition:   DNSNoDNSPodsAvailable,
+			status:      operatorv1.ConditionFalse,
+			gracePeriod: time.Minute * 5,
+		},
```

Comment:
Does having a grace period make sense for `DNSNoDNSPodsAvailable`?  If no pods are available, that means DNS service is completely unavailable.

Reply:
This presumes that enough will come up eventually.  Maybe a much longer grace period?  I'm okay with not using a gracePeriod on this condition, but this was the only other condition I chose because it seemed like it could easily change from no DNS pods available to less than the number of DNS pods available, which is the only other grace period condition.

Reply:
I removed the grace period for this.  We can add it later if it makes sense.